servers/rendering/shader_preprocessor.h

```diff
@@ -0,0 +1,202 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef SHADER_PREPROCESSOR_H
+#define SHADER_PREPROCESSOR_H
+
+#include "core/string/ustring.h"
+#include "core/templates/list.h"
+#include "core/templates/local_vector.h"
+#include "core/templates/rb_map.h"
+#include "core/templates/rb_set.h"
+#include "core/typedefs.h"
+
+#include "core/io/resource_loader.h"
+#include "core/os/os.h"
+#include "scene/resources/shader.h"
+#include "scene/resources/shader_include.h"
+
+class ShaderPreprocessor {
+public:
+	enum CompletionType {
+		COMPLETION_TYPE_NONE,
+		COMPLETION_TYPE_DIRECTIVE,
+		COMPLETION_TYPE_PRAGMA_DIRECTIVE,
+		COMPLETION_TYPE_PRAGMA,
+		COMPLETION_TYPE_INCLUDE_PATH,
+	};
+
+	struct FilePosition {
+		String file;
+		int line = 0;
+	};
+
+private:
+	struct Token {
+		char32_t text;
+		int line;
+
+		Token();
+		Token(char32_t p_text, int p_line);
+	};
+
+	// The real preprocessor that understands basic shader and preprocessor language syntax.
+	class Tokenizer {
+	public:
+		String code;
+		int line;
+		int index;
+		int size;
+		Vector<Token> generated;
+
+	private:
+		void add_generated(const Token &p_t);
+		char32_t next();
+
+	public:
+		int get_line() const;
+		int get_index() const;
+		char32_t peek();
+
+		void get_and_clear_generated(Vector<Token> *r_out);
+		void backtrack(char32_t p_what);
+		LocalVector<Token> advance(char32_t p_what);
+		void skip_whitespace();
+		String get_identifier(bool *r_is_cursor = nullptr, bool p_started = false);
+		String peek_identifier();
+		Token get_token();
+
+		Tokenizer(const String &p_code);
+	};
+
+	class CommentRemover {
+	private:
+		LocalVector<char32_t> stripped;
+		String code;
+		int index;
+		int line;
+		int comment_line_open;
+		int comments_open;
+		int strings_open;
+
+	public:
+		String get_error() const;
+		int get_error_line() const;
+		char32_t peek() const;
+
+		bool advance(char32_t p_what);
+		String strip();
+
+		CommentRemover(const String &p_code);
+	};
+
+	struct Define {
+		Vector<String> arguments;
+		String body;
+	};
+
+	struct SkippedCondition {
+		int start_line = -1;
+		int end_line = -1;
+	};
+
+	struct State {
+		RBMap<String, Define *> defines;
+		Vector<bool> skip_stack_else;
+		int condition_depth = 0;
+		RBSet<String> includes;
+		List<uint64_t> cyclic_include_hashes; // Holds code hash of includes.
+		int include_depth = 0;
+		String current_include;
+		String current_shader_type;
+		int shader_type_pos = -1;
```

Comment:
Remove `shader_type_pos` - if we don't need handling of it (to prevent preprocessor directive declaration before shader type.